### PR TITLE
[SYCL] IVDep: Add support for struct members and pointers

### PR DESCRIPTION
**Pull request by erichkeane**
Struct members was missed in the initial implementation, we just need to
properly handle the MemberExpr when doing our SEMA (+ consider it in a
few other places).

Pointers should be allowed, since the arrays decay to pointer pretty
easily in the type system. The implementation is as simple as allowing
it in Sema, then making sure we pass the Decl to the GEP creation
function.

### CODE REVIEW

File: clang/lib/Sema/SemaStmtAttr.cpp [hunk 1]
```diff
@@ -148,8 +148,8 @@ static IVDepExprResult HandleFPGAIVDepAttrExpr(Sema &S, Expr *E,
     return IVDepExprResult::SafeLen;
   }
 
-  if (isa<DeclRefExpr>(E)) {
-    if (!cast<DeclRefExpr>(E)->getType()->isArrayType()) {
+  if (isa<DeclRefExpr>(E) | isa<MemberExpr>(E)) {
```
Comment (premanandrao) [inline clang/lib/Sema/SemaStmtAttr.cpp:151]: I think you want the '||' operator here

File: clang/test/CodeGenSYCL/intel-fpga-ivdep-array.cpp [hunk 2]
```diff
@@ -225,3 +259,18 @@ int main() {
 // CHECK-DAG: ![[IVDEP_A_MUL_ARR_AND_GLOB]] = !{!"llvm.loop.parallel_access_indices", ![[IDX_GROUP_A_MUL_ARR_AND_GLOB]], i32 5}
 // CHECK-DAG: ![[IVDEP_B_MUL_ARR_AND_GLOB]] = !{!"llvm.loop.parallel_access_indices", ![[IDX_GROUP_B_MUL_ARR_AND_GLOB]], i32 6}
 // CHECK-DAG: ![[IVDEP_C_MUL_ARR_AND_GLOB]] = !{!"llvm.loop.parallel_access_indices", ![[IDX_GROUP_C_MUL_ARR_AND_GLOB]]}
+
+// CHECK-DAG: ![[IDX_GROUP_PTR]] = distinct !{}
+// CHECK-DAG: ![[MD_LOOP_PTR]] = distinct !{![[MD_LOOP_PTR]], ![[IVDEP_PTR:[0-9]+]]}
+// CHECK-DAG: ![[IVDEP_PTR]] = !{!"llvm.loop.parallel_access_indices", ![[IDX_GROUP_PTR]], i32 5}
+// TODO: ERICH: FIGURE OUT THESE VALUES
```
Comment (premanandrao) [inline clang/test/CodeGenSYCL/intel-fpga-ivdep-array.cpp:266]: Did you mean to leave the TODOs still?

File: clang/lib/Sema/SemaStmtAttr.cpp (hunk 1 above)
Comment (erichkeane) [reply clang/lib/Sema/SemaStmtAttr.cpp]: Yikes!  Thanks!

File: clang/test/CodeGenSYCL/intel-fpga-ivdep-array.cpp (hunk 2 above)
Comment (erichkeane) [reply clang/test/CodeGenSYCL/intel-fpga-ivdep-array.cpp]: No, good catch.  Apparently it got caught up in my copy/paste/update cycle :)

File: clang/lib/CodeGen/CGExpr.cpp [hunk 3]
```diff
@@ -3783,10 +3783,19 @@ LValue CodeGenFunction::EmitArraySubscriptExpr(const ArraySubscriptExpr *E,
...
+      PtrDecl = DRE->getDecl();
+    else if (const auto *ME =
+                 dyn_cast<MemberExpr>(E->getBase()->IgnoreParenCasts()))
+      PtrDecl = ME->getMemberDecl();
+
     Addr = emitArraySubscriptGEP(*this, Addr, Idx, E->getType(),
                                  !getLangOpts().isSignedOverflowDefined(),
                                  SignedIndices, E->getExprLoc(), &ptrType,
-                                 E->getBase());
+                                 E->getBase(), "ptridx", PtrDecl);
```
Comment (premanandrao) [inline clang/lib/CodeGen/CGExpr.cpp:3798]: Is this name expected to appear in the output?  Otherwise, LGTM.

Comment (erichkeane) [reply clang/lib/CodeGen/CGExpr.cpp:3798]: Yes, its basically the 'hint' for the name of the GEP in IR.  That said, those are often stripped in non-debug builds :)

File: clang/include/clang/Basic/Attr.td [hunk 4]
```diff
@@ -2582,8 +2586,8 @@ def NoDeref : TypeAttr {
 // intel::reqd_work_group_size spelling.
 def ReqdWorkGroupSize : InheritableAttr {
   let Spellings = [GNU<"reqd_work_group_size">,
-                   CXX11<"intel","reqd_work_group_size">, 
-                   CXX11<"cl","reqd_work_group_size">]; 
+                   CXX11<"intel","reqd_work_group_size">,
+                   CXX11<"cl","reqd_work_group_size">];
```
Comment (Fznamznon) [inline clang/include/clang/Basic/Attr.td:2590]: Seems like unrelated change.

Comment (erichkeane) [reply clang/include/clang/Basic/Attr.td:2590]: Ah, shoot... my editor is configured to remove trailing WS on save and must have picked it up.  @bader, you resolved this conversation, are you OK with it, or would you like me to revert this?

**Comment by AGindinson** [comment]
Nice one, thanks Erich!
Taking this further, is there a way to allow all `operator[]` implementers (like `cl::sycl::accessor`) implementers into the attribute's arguments? Or am I missing something and it's already facilitated by the pointer-supporting code?

**Comment by erichkeane** [comment]
> 
> 
> Nice one, thanks Erich!
> Taking this further, is there a way to allow all `operator[]` implementers (like `cl::sycl::accessor`) implementers into the attribute's arguments? Or am I missing something and it's already facilitated by the pointer-supporting code?

There isn't really.  If an attribute isn't something that modifies the type, the source code itself has no way to get to the attribute.  What you CAN do is make the attribute and the code both get the argument the same way, such as a single constant expression that they both reference.

File: clang/include/clang/Basic/Attr.td (hunk 4 above)
Comment (bader) [reply clang/include/clang/Basic/Attr.td:2590]: I'm okay with it.
